### PR TITLE
Reapply session_base_t refactoring

### PR DESCRIPTION
**Pull request by sigiesec**
Re-apply the changes from  6d7aeb0 and fix Sun Studio compilation issues.

### CODE REVIEW

**Comment by sigiesec** [comment]
@bluca Can you check if this works with Sun Studio?

**Comment by bluca** [comment]
Thanks, I'll check this later tonight and let you know

**Comment by bluca** [comment]
Computer says no:

```
gmake[1]: Entering directory `/tmp/zeromq-4.3.0'
  CXX      src/src_libzmq_la-session_base.lo
"src/stream_engine.hpp", line 245: Warning: zmq::stream_engine_t::plug hides the function zmq::io_object_t::plug(zmq::io_thread_t*).
"src/udp_engine.hpp", line 69: Warning: zmq::udp_engine_t::plug hides the function zmq::io_object_t::plug(zmq::io_thread_t*).
"src/session_base.cpp", line 573: Error: Could not find a match for std::map<std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool), std::less<std::string>, std::allocator<std::pair<const std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool)>>>::map(std::pair<std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool)>[2], std::pair<std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool)>*) needed in<no tag>.
"/opt/developerstudio12.6/lib/compilers/include/CC/Cstd/map", line 144: Note: Candidate 'std::map<std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool), std::less<std::string>, std::allocator<std::pair<const std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool)>>>::map(const std::less<std::string>&, const std::allocator<std::pair<const std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool)>>&)' is not viable: argument 'comp' can't be converted from 'std::pair<std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool)>[2]' to 'const std::less<std::string>&'.
"/opt/developerstudio12.6/lib/compilers/include/CC/Cstd/map", line 195: Note: Candidate 'std::map<std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool), std::less<std::string>, std::allocator<std::pair<const std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool)>>>::map(const std::map<std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool), std::less<std::string>, std::allocator<std::pair<const std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool)>>>&)' is not viable: too many arguments.
"/opt/developerstudio12.6/lib/compilers/include/CC/Cstd/rw/tree", line 393: Note: Candidate '__rwstd::__rb_tree<std::string, std::pair<const std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool)>, __rwstd::__select1st<std::pair<const std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool)>, std::string>, std::less<std::string>, std::allocator<std::pair<const std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool)>>>::const_iterator::const_iterator()' is not viable: too many arguments.
"/opt/developerstudio12.6/lib/compilers/include/CC/Cstd/map", line 171: Note: Candidate 'std::map<std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool), std::less<std::string>, std::allocator<std::pair<const std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool)>>>::map(__rwstd::__rb_tree<std::string, std::pair<const std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool)>, __rwstd::__select1st<std::pair<const std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool)>, std::string>, std::less<std::string>, std::allocator<std::pair<const std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool)>>>::const_iterator, __rwstd::__rb_tree<std::string, std::pair<const std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool)>, __rwstd::__select1st<std::pair<const std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool)>, std::string>, std::less<std::string>, std::allocator<std::pair<const std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool)>>>::const_iterator, const std::less<std::string>&, const std::allocator<std::pair<const std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool)>>&)' is not viable: argument 'first' can't be converted from 'std::pair<std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool)>[2]' to '__rwstd::__rb_tree<std::string, std::pair<const std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool)>, __rwstd::__select1st<std::pair<const std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool)>, std::string>, std::less<std::string>, std::allocator<std::pair<const std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool)>>>::const_iterator'.
"/opt/developerstudio12.6/lib/compilers/include/CC/Cstd/rw/tree", line 320: Note: Candidate '__rwstd::__rb_tree<std::string, std::pair<const std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool)>, __rwstd::__select1st<std::pair<const std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool)>, std::string>, std::less<std::string>, std::allocator<std::pair<const std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool)>>>::iterator::iterator()' is not viable: too many arguments.
"/opt/developerstudio12.6/lib/compilers/include/CC/Cstd/map", line 167: Note: Candidate 'std::map<std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool), std::less<std::string>, std::allocator<std::pair<const std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool)>>>::map(__rwstd::__rb_tree<std::string, std::pair<const std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool)>, __rwstd::__select1st<std::pair<const std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool)>, std::string>, std::less<std::string>, std::allocator<std::pair<const std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool)>>>::iterator, __rwstd::__rb_tree<std::string, std::pair<const std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool)>, __rwstd::__select1st<std::pair<const std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool)>, std::string>, std::less<std::string>, std::allocator<std::pair<const std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool)>>>::iterator, const std::less<std::string>&, const std::allocator<std::pair<const std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool)>>&)' is not viable: argument 'first' can't be converted from 'std::pair<std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool)>[2]' to '__rwstd::__rb_tree<std::string, std::pair<const std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool)>, __rwstd::__select1st<std::pair<const std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool)>, std::string>, std::less<std::string>, std::allocator<std::pair<const std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool)>>>::iterator'.
"/opt/developerstudio12.6/lib/compilers/include/CC/Cstd/map", line 163: Note: Candidate 'std::map<std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool), std::less<std::string>, std::allocator<std::pair<const std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool)>>>::map(const std::pair<const std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool)>*, const std::pair<const std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool)>*, const std::less<std::string>&, const std::allocator<std::pair<const std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool)>>&)' is not viable: argument 'first' can't be converted from 'std::pair<std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool)>[2]' to 'const std::pair<const std::string, zmq::own_t*(zmq::session_base_t::*)(zmq::io_thread_t*,bool)>*'.
"src/session_base.cpp", line 595: Error: Could not find a match for std::map<std::string,void(zmq::session_base_t::*)(zmq::io_thread_t*)>::map(std::pair<std::string, void(zmq::session_base_t::*)(zmq::io_thread_t*)>[1], std::pair<std::string, void(zmq::session_base_t::*)(zmq::io_thread_t*)>*) needed in<no tag>.
"/opt/developerstudio12.6/lib/compilers/include/CC/Cstd/map", line 144: Note: Candidate 'std::map<std::string,void(zmq::session_base_t::*)(zmq::io_thread_t*)>::map(const std::less<std::string>&, const std::allocator<std::pair<const std::string, void(zmq::session_base_t::*)(zmq::io_thread_t*)>>&)' is not viable: argument 'comp' can't be converted from 'std::pair<std::string, void(zmq::session_base_t::*)(zmq::io_thread_t*)>[1]' to 'const std::less<std::string>&'.
"/opt/developerstudio12.6/lib/compilers/include/CC/Cstd/map", line 195: Note: Candidate 'std::map<std::string,void(zmq::session_base_t::*)(zmq::io_thread_t*)>::map(const std::map<std::string,void(zmq::session_base_t::*)(zmq::io_thread_t*)>&)' is not viable: too many arguments.
"/opt/developerstudio12.6/lib/compilers/include/CC/Cstd/rw/tree", line 393: Note: Candidate '__rwstd::__rb_tree<std::string, std::pair<const std::string, void(zmq::session_base_t::*)(zmq::io_thread_t*)>, __rwstd::__select1st<std::pair<const std::string, void(zmq::session_base_t::*)(zmq::io_thread_t*)>, std::string>, std::less<std::string>, std::allocator<std::pair<const std::string, void(zmq::session_base_t::*)(zmq::io_thread_t*)>>>::const_iterator::const_iterator()' is not viable: too many arguments.
"/opt/developerstudio12.6/lib/compilers/include/CC/Cstd/map", line 171: Note: Candidate 'std::map<std::string,void(zmq::session_base_t::*)(zmq::io_thread_t*)>::map(__rwstd::__rb_tree<std::string, std::pair<const std::string, void(zmq::session_base_t::*)(zmq::io_thread_t*)>, __rwstd::__select1st<std::pair<const std::string, void(zmq::session_base_t::*)(zmq::io_thread_t*)>, std::string>, std::less<std::string>, std::allocator<std::pair<const std::string, void(zmq::session_base_t::*)(zmq::io_thread_t*)>>>::const_iterator, __rwstd::__rb_tree<std::string, std::pair<const std::string, void(zmq::session_base_t::*)(zmq::io_thread_t*)>, __rwstd::__select1st<std::pair<const std::string, void(zmq::session_base_t::*)(zmq::io_thread_t*)>, std::string>, std::less<std::string>, std::allocator<std::pair<const std::string, void(zmq::session_base_t::*)(zmq::io_thread_t*)>>>::const_iterator, const std::less<std::string>&, const std::allocator<std::pair<const std::string, void(zmq::session_base_t::*)(zmq::io_thread_t*)>>&)' is not viable: argument 'first' can't be converted from 'std::pair<std::string, void(zmq::session_base_t::*)(zmq::io_thread_t*)>[1]' to '__rwstd::__rb_tree<std::string, std::pair<const std::string, void(zmq::session_base_t::*)(zmq::io_thread_t*)>, __rwstd::__select1st<std::pair<const std::string, void(zmq::session_base_t::*)(zmq::io_thread_t*)>, std::string>, std::less<std::string>, std::allocator<std::pair<const std::string, void(zmq::session_base_t::*)(zmq::io_thread_t*)>>>::const_iterator'.
"/opt/developerstudio12.6/lib/compilers/include/CC/Cstd/rw/tree", line 320: Note: Candidate '__rwstd::__rb_tree<std::string, std::pair<const std::string, void(zmq::session_base_t::*)(zmq::io_thread_t*)>, __rwstd::__select1st<std::pair<const std::string, void(zmq::session_base_t::*)(zmq::io_thread_t*)>, std::string>, std::less<std::string>, std::allocator<std::pair<const std::string, void(zmq::session_base_t::*)(zmq::io_thread_t*)>>>::iterator::iterator()' is not viable: too many arguments.
"/opt/developerstudio12.6/lib/compilers/include/CC/Cstd/map", line 167: Note: Candidate 'std::map<std::string,void(zmq::session_base_t::*)(zmq::io_thread_t*)>::map(__rwstd::__rb_tree<std::string, std::pair<const std::string, void(zmq::session_base_t::*)(zmq::io_thread_t*)>, __rwstd::__select1st<std::pair<const std::string, void(zmq::session_base_t::*)(zmq::io_thread_t*)>, std::string>, std::less<std::string>, std::allocator<std::pair<const std::string, void(zmq::session_base_t::*)(zmq::io_thread_t*)>>>::iterator, __rwstd::__rb_tree<std::string, std::pair<const std::string, void(zmq::session_base_t::*)(zmq::io_thread_t*)>, __rwstd::__select1st<std::pair<const std::string, void(zmq::session_base_t::*)(zmq::io_thread_t*)>, std::string>, std::less<std::string>, std::allocator<std::pair<const std::string, void(zmq::session_base_t::*)(zmq::io_thread_t*)>>>::iterator, const std::less<std::string>&, const std::allocator<std::pair<const std::string, void(zmq::session_base_t::*)(zmq::io_thread_t*)>>&)' is not viable: argument 'first' can't be converted from 'std::pair<std::string, void(zmq::session_base_t::*)(zmq::io_thread_t*)>[1]' to '__rwstd::__rb_tree<std::string, std::pair<const std::string, void(zmq::session_base_t::*)(zmq::io_thread_t*)>, __rwstd::__select1st<std::pair<const std::string, void(zmq::session_base_t::*)(zmq::io_thread_t*)>, std::string>, std::less<std::string>, std::allocator<std::pair<const std::string, void(zmq::session_base_t::*)(zmq::io_thread_t*)>>>::iterator'.
"/opt/developerstudio12.6/lib/compilers/include/CC/Cstd/map", line 163: Note: Candidate 'std::map<std::string,void(zmq::session_base_t::*)(zmq::io_thread_t*)>::map(const std::pair<const std::string, void(zmq::session_base_t::*)(zmq::io_thread_t*)>*, const std::pair<const std::string, void(zmq::session_base_t::*)(zmq::io_thread_t*)>*, const std::less<std::string>&, const std::allocator<std::pair<const std::string, void(zmq::session_base_t::*)(zmq::io_thread_t*)>>&)' is not viable: argument 'first' can't be converted from 'std::pair<std::string, void(zmq::session_base_t::*)(zmq::io_thread_t*)>[1]' to 'const std::pair<const std::string, void(zmq::session_base_t::*)(zmq::io_thread_t*)>*'.
2 Error(s) and 2 Warning(s) detected.
```

It's the same error I got when switching to std::string(protocol:foo) with the previous version

**Comment by sigiesec** [comment]
Probably, the const was stil missing on the pair key type. Could you try again? Resp. can I see this on some CI server myself?

**Comment by bluca** [comment]
Thanks, I'll try again tonight (sorry didn't have time to mess with it myself yesterday)

> Resp. can I see this on some CI server myself?

I wish :-( I just have an x86 Solaris 10 VM (which is the last version you can get for free), where I somehow managed to install Sun Studio 12.6. I should probably back that up somewhere...

**Comment by bluca** [comment]
This works, thanks!

File: src/session_base.cpp [hunk 1]
```diff
@@ -550,6 +550,54 @@ void zmq::session_base_t::reconnect ()
...
+                               &zmq::session_base_t::create_connecter_tipc),
+#endif
+#if defined ZMQ_HAVE_VMCI
+    connecter_factory_entry_t (protocol_name::vmci,
+                               &zmq::session_base_t::create_connecter_vmci),
+#endif
+};
+
+zmq::session_base_t::connecter_factory_map_t
+  zmq::session_base_t::_connecter_factories_map (
```
Comment (bluca) [inline src/session_base.cpp:573]: @sigiesec address sanitizer is complaining about uninitialized value here (line numbers have changed), any idea?

https://travis-ci.org/github/google/oss-fuzz/jobs/680325364

```
  Uninitialized value was created by a heap allocation
    #0 0x525d89 in operator new(unsigned long) /src/llvm-project/compiler-rt/lib/msan/msan_new_delete.cpp:45:35
    #1 0x741319 in std::__1::unique_ptr<std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)>, void*>, std::__1::__tree_node_destructor<std::__1::allocator<std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)>, void*> > > > std::__1::__tree<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)>, std::__1::__map_value_compare<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)>, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, true>, std::__1::allocator<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)> > >::__construct_node<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)> const&>(std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)> const&) /usr/local/bin/../include/c++/v1/__tree:2190:23
    #2 0x7402c8 in std::__1::__tree_iterator<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)>, std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)>, void*>*, long> std::__1::__tree<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)>, std::__1::__map_value_compare<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)>, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, true>, std::__1::allocator<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)> > >::__emplace_hint_unique_key_args<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)> const&>(std::__1::__tree_const_iterator<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)>, std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)>, void*>*, long>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)> const&) /usr/local/bin/../include/c++/v1/__tree:2169:29
    #3 0x73fa3d in std::__1::__map_iterator<std::__1::__tree_iterator<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)>, std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)>, void*>*, long> > std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool), std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)> > >::insert<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)>&, void>(std::__1::__map_const_iterator<std::__1::__tree_const_iterator<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)>, std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)>, void*>*, long> >, std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)>&) /usr/local/bin/../include/c++/v1/map:1167:29
    #4 0x73f46c in void std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool), std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)> > >::insert<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)>*>(std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)>*, std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)>*) /usr/local/bin/../include/c++/v1/map:1199:17
    #5 0x73cfe1 in std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool), std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)> > >::map<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)>*>(std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)>*, std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, zmq::own_t* (zmq::session_base_t::*)(zmq::io_thread_t*, bool)>*, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > const&) /usr/local/bin/../include/c++/v1/map:984:13
    #6 0x41f037 in __cxx_global_var_init.23 /src/libzmq/src/session_base.cpp:626:24
    #7 0x41f2a5 in _GLOBAL__sub_I_session_base.cpp /src/libzmq/src/session_base.cpp
    #8 0x8d0f6c in __libc_csu_init (/tmp/not-out/test_connect_null_fuzzer+0x8d0f6c)
SUMMARY: MemorySanitizer: use-of-uninitialized-value /usr/local/bin/../include/c++/v1/string:947:17 in std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::size() const
```

Comment (bluca) [reply src/session_base.cpp:573]: Reverted this commit via https://github.com/zeromq/libzmq/pull/3896

**Pull request merged**